File: tests/fixtures/sample_ts_app/auth_service.ts
// --- SIMPLE FUNCTIONS ---

function greet(name: string): string {
    return `Hello, ${name}!`;
}

function add(a: number, b: number): number {
    return a + b;
}

// --- ASYNC FUNCTION ---

async function fetchUserProfile(userId: string): Promise<User> {
    const response = await fetch(`/api/users/${userId}`);
    if (!response.ok) {
        throw new Error(`Failed to fetch user: ${response.status}`);
    }
    return response.json();
}

// --- INTERFACE ---

interface User {
    id: string;
    name: string;
    email: string;
    role: "admin" | "user" | "moderator";
}

// --- CLASS WITH METHODS ---

class AuthService {
    private tokenKey = "auth_token";
    private baseUrl: string;

    constructor(baseUrl: string) {
        this.baseUrl = baseUrl;
    }

    async login(email: string, password: string): Promise<string> {
        const response = await fetch(`${this.baseUrl}/auth/login`, {
            method: "POST",
            headers: { "Content-Type": "application/json" },
            body: JSON.stringify({ email, password }),
        });

        if (!response.ok) {
            throw new Error("Authentication failed");
        }

        const data = await response.json();
        localStorage.setItem(this.tokenKey, data.token);
        return data.token;
    }

    logout(): void {
        localStorage.removeItem(this.tokenKey);
    }

    isAuthenticated(): boolean {
        return localStorage.getItem(this.tokenKey) !== null;
    }

    getToken(): string | null {
        return localStorage.getItem(this.tokenKey);
    }
}

// --- GENERIC FUNCTION ---

function filterByProperty<T>(items: T[], key: keyof T, value: T[keyof T]): T[] {
    return items.filter((item) => item[key] === value);
}

// --- FUNCTION WITH DESTRUCTURED PARAMS ---

function createUser({ name, email, role }: { name: string; email: string; role: string }): User {
    return {
        id: crypto.randomUUID(),
        name,
        email,
        role: role as User["role"],
    };
}

// --- CLASS WITH INHERITANCE ---

class AdminService extends AuthService {
    private adminEndpoint: string;

    constructor(baseUrl: string, adminEndpoint: string) {
        super(baseUrl);
        this.adminEndpoint = adminEndpoint;
    }

    async getAllUsers(): Promise<User[]> {
        const token = this.getToken();
        if (!token) {
            throw new Error("Not authenticated");
        }

        const response = await fetch(`${this.adminEndpoint}/users`, {
            headers: { Authorization: `Bearer ${token}` },
        });

        return response.json();
    }

    async deleteUser(userId: string): Promise<void> {
        const token = this.getToken();
        await fetch(`${this.adminEndpoint}/users/${userId}`, {
            method: "DELETE",
            headers: { Authorization: `Bearer ${token}` },
        });
    }
}

File: tests/fixtures/sample_dart_app/auth_service.dart
import 'dart:async';
import 'dart:convert';

// --- SIMPLE FUNCTIONS ---

String greet(String name) {
  return 'Hello, $name!';
}

int add(int a, int b) {
  return a + b;
}

// --- ASYNC FUNCTION ---

Future<Map<String, dynamic>> fetchUserProfile(String userId) async {
  final response = await http.get(Uri.parse('/api/users/$userId'));
  if (response.statusCode != 200) {
    throw Exception('Failed to fetch user: ${response.statusCode}');
  }
  return jsonDecode(response.body);
}

// --- CLASS WITH METHODS ---

class AuthService {
  final String baseUrl;
  String? _token;

  AuthService(this.baseUrl);

  // Named constructor
  AuthService.withToken(this.baseUrl, this._token);

  Future<String> login(String email, String password) async {
    final response = await http.post(
      Uri.parse('$baseUrl/auth/login'),
      headers: {'Content-Type': 'application/json'},
      body: jsonEncode({'email': email, 'password': password}),
    );

    if (response.statusCode != 200) {
      throw Exception('Authentication failed');
    }

    final data = jsonDecode(response.body);
    _token = data['token'];
    return _token!;
  }

  void logout() {
    _token = null;
  }

  bool get isAuthenticated => _token != null;

  String? getToken() {
    return _token;
  }
}

// --- ABSTRACT CLASS ---

abstract class Repository<T> {
  Future<T> findById(String id);
  Future<List<T>> findAll();
  Future<void> save(T entity);
  Future<void> delete(String id);
}

// --- CLASS WITH INHERITANCE + MIXINS ---

mixin Loggable {
  void log(String message) {
    print('[${DateTime.now()}] $message');
  }
}

class UserRepository extends Repository<Map<String, dynamic>> with Loggable {
  final String apiUrl;
  final AuthService authService;

  UserRepository(this.apiUrl, this.authService);

  @override
  Future<Map<String, dynamic>> findById(String id) async {
    log('Fetching user: $id');
    final token = authService.getToken();
    final response = await http.get(
      Uri.parse('$apiUrl/users/$id'),
      headers: {'Authorization': 'Bearer $token'},
    );
    return jsonDecode(response.body);
  }

  @override
  Future<List<Map<String, dynamic>>> findAll() async {
    log('Fetching all users');
    final token = authService.getToken();
    final response = await http.get(
      Uri.parse('$apiUrl/users'),
      headers: {'Authorization': 'Bearer $token'},
    );
    final List<dynamic> data = jsonDecode(response.body);
    return data.cast<Map<String, dynamic>>();
  }

  @override
  Future<void> save(Map<String, dynamic> entity) async {
    log('Saving user: ${entity['id']}');
    final token = authService.getToken();
    await http.post(
      Uri.parse('$apiUrl/users'),
      headers: {
        'Authorization': 'Bearer $token',
        'Content-Type': 'application/json',
      },
      body: jsonEncode(entity),
    );
  }

  @override
  Future<void> delete(String id) async {
    log('Deleting user: $id');
    final token = authService.getToken();
    await http.delete(
      Uri.parse('$apiUrl/users/$id'),
      headers: {'Authorization': 'Bearer $token'},
    );
  }
}

// --- FUNCTION WITH NAMED PARAMETERS ---

Map<String, dynamic> createUser({
  required String name,
  required String email,
  String role = 'user',
  bool isActive = true,
}) {
  return {
    'id': DateTime.now().millisecondsSinceEpoch.toString(),
    'name': name,
    'email': email,
    'role': role,
    'isActive': isActive,
  };
}

// --- ENUM ---

enum UserRole {
  admin,
  moderator,
  user;

  String get displayName {
    switch (this) {
      case UserRole.admin:
        return 'Administrator';
      case UserRole.moderator:
        return 'Moderator';
      case UserRole.user:
        return 'Regular User';
    }
  }
}
